File: v4/rust/iprange-livedb/src/publication/problem.rs
//! Fixed, allocation-free publication failure details.

use crate::error::{Error as SdkError, ErrorCode};

use super::main_file;
use super::namespace::NamespaceError;
use super::output;
use super::replacement;
use super::reservation_file;
pub(crate) use super::types::PublicationProblem as Problem;

impl Problem {
    #[cfg(all(test, unix))]
    pub(crate) const fn injected() -> Self {
        Self::new(ErrorCode::Io, None, "injected publication failure")
    }

    pub(crate) const fn cleanup_conflict(detail: &'static str) -> Self {
        Self::new(ErrorCode::CleanupConflict, None, detail)
    }

    #[cfg(windows)]
    pub(crate) const fn cleanup_in_progress(detail: &'static str) -> Self {
        Self::new(ErrorCode::CleanupInProgress, None, detail)
    }

    pub(crate) fn namespace(error: &NamespaceError) -> Self {
        match error {
            NamespaceError::InvalidName => {
                Self::plain(ErrorCode::NameInvalid, "invalid destination name")
            }
            NamespaceError::NotDirectory => {
                Self::plain(ErrorCode::Conflict, "destination parent is not a directory")
            }
            NamespaceError::NotRegular => Self::plain(
                ErrorCode::Conflict,
                "publication name is not a regular file",
            ),
            NamespaceError::Exists => {
                Self::plain(ErrorCode::NameExists, "publication name already exists")
            }
            NamespaceError::Missing => {
                Self::plain(ErrorCode::NameNotFound, "publication name is missing")
            }
            NamespaceError::IdentityChanged => {
                Self::plain(ErrorCode::Conflict, "publication inode identity changed")
            }
            NamespaceError::LinkCount(links) if *links == 0 => {
                Self::plain(ErrorCode::Conflict, "publication inode has no links")
            }
            NamespaceError::LinkCount(_) => {
                Self::plain(ErrorCode::Conflict, "publication inode link count changed")
            }
            NamespaceError::CrossFilesystem => Self::plain(
                ErrorCode::PublicationUnsupported,
                "publication inode is on another filesystem",
            ),
            NamespaceError::AccessPolicy => Self::plain(
                ErrorCode::AccessPolicyUnsupported,
                "creator-only access policy is not proved",
            ),
            NamespaceError::Unsupported => Self::plain(
                ErrorCode::DurabilityUnsupported,
                "filesystem lacks required durable namespace operations",
            ),
            NamespaceError::ForkedHandle => {
                Self::plain(ErrorCode::ForkedHandle, "publication handle crossed fork")
            }
            NamespaceError::Io(source) => {
                Self::io(source, "publication filesystem operation failed")
            }
            NamespaceError::IoAt { source, .. } if source.raw_os_error() == Some(libc::ELOOP) => {
                Self::plain(ErrorCode::Conflict, "publication name is a symlink")
            }
            NamespaceError::IoAt { operation, source } => Self::io(source, operation),
        }
    }

    pub(crate) fn output(error: &output::Error) -> Self {
        match error {
            output::Error::Namespace(cause) => Self::namespace(cause),
            output::Error::Sdk(cause) => Self::sdk(cause),
            output::Error::Bootstrap => {
                Self::plain(ErrorCode::FormatInvalid, "output metadata is malformed")
            }
            output::Error::Gc(problem) => problem.clone(),
            output::Error::FinishedMetaChanged => {
                Self::plain(ErrorCode::Conflict, "finished output metadata changed")
            }
            output::Error::FinishedLengthChanged => {
                Self::plain(ErrorCode::Conflict, "finished output length changed")
            }
        }
    }

    pub(crate) fn reservation(error: &reservation_file::Error) -> Self {
        match error {
            reservation_file::Error::Namespace(cause) => Self::namespace(cause),
            reservation_file::Error::Sdk(cause) => Self::sdk(cause),
            reservation_file::Error::Output(cause) => Self::output(cause),
            reservation_file::Error::Gc(problem) => problem.clone(),
            reservation_file::Error::Checkpoint(problem) => problem.clone(),
            reservation_file::Error::Codec => {
                Self::plain(ErrorCode::FormatInvalid, "reservation record is malformed")
            }
            reservation_file::Error::HeaderChanged => {
                Self::plain(ErrorCode::Conflict, "reservation record changed")
            }
            reservation_file::Error::HeaderInvariant => Self::plain(
                ErrorCode::FormatInvalid,
                "reservation state is inconsistent",
            ),
            reservation_file::Error::LengthChanged => {
                Self::plain(ErrorCode::Conflict, "reservation length changed")
            }
        }
    }

    pub(crate) fn replacement(error: &replacement::Error) -> Self {
        match error {
            replacement::Error::Namespace(cause) => Self::namespace(cause),
            replacement::Error::Sdk(cause) => Self::sdk(cause),
            replacement::Error::Output(cause) => Self::output(cause),
            replacement::Error::SameIdentity => Self::plain(
                ErrorCode::Conflict,
                "replacement source and destination identities match",
            ),
            replacement::Error::ContentChanged => Self::plain(
                ErrorCode::Conflict,
                "replacement destination content changed",
            ),
        }
    }

    pub(crate) fn main(error: &main_file::Error) -> Self {
        match error {
            main_file::Error::Namespace(cause) => Self::namespace(cause),
            main_file::Error::Sdk(cause) => Self::sdk(cause),
            main_file::Error::Output(cause) => Self::output(cause),
            main_file::Error::Reservation(cause) => Self::reservation(cause),
            main_file::Error::Checkpoint(problem) => problem.clone(),
            #[cfg(unix)]
            main_file::Error::PreviousLinkCount => Self::plain(
                ErrorCode::CleanupConflict,
                "retired previous destination still has a link",
            ),
            #[cfg(unix)]
            main_file::Error::ReservationLinkCount => Self::plain(
                ErrorCode::CleanupConflict,
                "retired reservation still has a link",
            ),
            #[cfg(windows)]
            main_file::Error::Gc(problem) => problem.clone(),
            #[cfg(all(test, unix))]
            main_file::Error::Injected => Self::injected(),
        }
    }

    const fn plain(code: ErrorCode, detail: &'static str) -> Self {
        Self::new(code, None, detail)
    }

    fn io(source: &std::io::Error, detail: &'static str) -> Self {
        Self::new(ErrorCode::Io, source.raw_os_error(), detail)
    }

    pub(crate) fn sdk(error: &SdkError) -> Self {
        let os_code = match error {
            SdkError::Io(source) => source.raw_os_error(),
            _ => None,
        };
        Self::new(error.code(), os_code, "publication SDK operation failed")
    }

    #[cfg(windows)]
    pub(crate) fn into_sdk(self) -> SdkError {
        let code = self.code;
        let os_code = self.os_code;
        let detail = match self.detail {
            std::borrow::Cow::Borrowed(detail) => Some(detail),
            std::borrow::Cow::Owned(_) => None,
        };
        match (code, os_code, detail) {
            (ErrorCode::Io, Some(code), _) => SdkError::Io(std::io::Error::from_raw_os_error(code)),
            (ErrorCode::NameInvalid, _, _) => SdkError::NameInvalid,
            (ErrorCode::NameExists, _, _) => SdkError::NameExists,
            (ErrorCode::NameNotFound, _, _) => SdkError::NameNotFound,
            (ErrorCode::DurabilityUnsupported, _, Some(detail)) => {
                SdkError::DurabilityUnsupported(detail)
            }
            (ErrorCode::CleanupInProgress, _, Some(detail)) => SdkError::CleanupInProgress(detail),
            (ErrorCode::CleanupConflict, _, Some(detail)) => SdkError::CleanupConflict(detail),
            (ErrorCode::DirectoryIdentityMismatch, _, _) => SdkError::DirectoryIdentityMismatch,
            (ErrorCode::ForkedHandle, _, _) => SdkError::ForkedHandle,
            (_, _, Some(detail)) => SdkError::Conflict(detail),
            _ => SdkError::WorkerOperation { code, os_code },
        }
    }
}

File: v4/go/internal/publication/problem.go
// Fixed, allocation-free publication failure details (Rust
// publication/problem.rs). Every code and detail string is verbatim
// Rust; Go does not carry os_code (design decision 6), and the
// Windows-only gc/checkpoint/in-progress arms are recorded with their
// Phase-2 and 4-10/4-11 chunks, never stubbed here.

package publication

import (
	"errors"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
)

// problem builds one fixed problem (Rust PublicationProblem::new).
func problem(code format.ErrorCode, detail string) *format.Error {
	return &format.Error{Code: code, Detail: detail}
}

// cleanupConflictProblem builds the fixed cleanup-conflict class (Rust
// Problem::cleanup_conflict).
func cleanupConflictProblem(detail string) *format.Error {
	return problem(format.CodeCleanupConflict, detail)
}

// checkpointProblem is the Go peer of the Rust Error::Checkpoint
// (reservation_file::Error / main_file::Error) and Result::Checkpoint
// arms: one observer or machine checkpoint problem that is already in
// its final fixed shape and must pass every composition fold
// unchanged, exactly like the Rust problem clone.
type checkpointProblem struct {
	problem *format.Error
}

func (e *checkpointProblem) Error() string { return e.problem.Error() }
func (e *checkpointProblem) Unwrap() error { return e.problem }

// asCheckpointProblem unwraps one checkpoint problem, or nil.
func asCheckpointProblem(err error) *format.Error {
	var checkpoint *checkpointProblem
	if errors.As(err, &checkpoint) {
		return checkpoint.problem
	}
	return nil
}

// namespaceProblem maps one retained-directory namespace error to the
// fixed publication problem (Rust Problem::namespace). The plain Io
// class always reports the fixed filesystem-operation detail; the
// IoAt class reports its operation label, and a no-follow final
// symlink folds to the Conflict class.
func namespaceProblem(err error) *format.Error {
	// The creator-only security owner reports the AccessPolicy class
	// directly (Rust folds it into NamespaceError::AccessPolicy at the
	// machine boundary; both fold to the same problem here).
	var fe *format.Error
	if errors.As(err, &fe) && fe.Code == format.CodeAccessPolicyUnsupported {
		return problem(format.CodeAccessPolicyUnsupported, "creator-only access policy is not proved")
	}
	nerr, ok := live.AsNamespaceError(err)
	if !ok {
		return sdkProblem(err)
	}
	switch nerr.Kind {
	case live.NamespaceInvalidName:
		return problem(format.CodeNameInvalid, "invalid destination name")
	case live.NamespaceNotDirectory:
		return problem(format.CodeConflict, "destination parent is not a directory")
	case live.NamespaceNotRegular:
		return problem(format.CodeConflict, "publication name is not a regular file")
	case live.NamespaceExists:
		return problem(format.CodeNameExists, "publication name already exists")
	case live.NamespaceMissing:
		return problem(format.CodeNameNotFound, "publication name is missing")
	case live.NamespaceIdentityChanged:
		return problem(format.CodeConflict, "publication inode identity changed")
	case live.NamespaceLinkCount:
		if nerr.Links == 0 {
			return problem(format.CodeConflict, "publication inode has no links")
		}
		return problem(format.CodeConflict, "publication inode link count changed")
	case live.NamespaceCrossFilesystem:
		return problem(format.CodePublicationUnsupported, "publication inode is on another filesystem")
	case live.NamespaceAccessPolicy:
		return problem(format.CodeAccessPolicyUnsupported, "creator-only access policy is not proved")
	case live.NamespaceUnsupported:
		return problem(format.CodeDurabilityUnsupported, "filesystem lacks required durable namespace operations")
	case live.NamespaceForkedHandle:
		return problem(format.CodeForkedHandle, "publication handle crossed fork")
	case live.NamespaceIo:
		return problem(format.CodeIO, "publication filesystem operation failed")
	case live.NamespaceIoAt:
		if live.IsNofollowSymlink(nerr.Err) {
			return problem(format.CodeConflict, "publication name is a symlink")
		}
		return problem(format.CodeIO, nerr.Op)
	}
	panic("unreachable namespace error kind")
}

// outputProblem maps one output-machine failure (Rust Problem::output:
// Namespace, Sdk, Bootstrap, Gc, FinishedMetaChanged,
// FinishedLengthChanged). The Gc arm is Windows-only and recorded with
// Phase-2, never stubbed here.
func outputProblem(err error) *format.Error {
	if _, ok := live.AsNamespaceError(err); ok {
		return namespaceProblem(err)
	}
	var fe *format.Error
	if !errors.As(err, &fe) {
		return sdkProblem(err)
	}
	switch fe.Code {
	case format.CodeAccessPolicyUnsupported:
		return problem(format.CodeAccessPolicyUnsupported, "creator-only access policy is not proved")
	case format.CodeFormatInvalid:
		if fe.Detail == "" {
			return problem(format.CodeFormatInvalid, "output metadata is malformed")
		}
		return fe
	case format.CodeConflict:
		// The finished-output arms carry Rust-verbatim fixed details
		// from the producer ("finished output metadata changed" /
		// "finished output length changed"); preserve them.
		return fe
	default:
		return sdkProblem(err)
	}
}

// reservationProblem maps one reservation-file failure (Rust
// Problem::reservation: Namespace, Sdk, Output, Gc, Checkpoint, Codec,
// HeaderChanged, HeaderInvariant, LengthChanged). The Gc and
// Checkpoint arms are Windows-only / 4-10-4-11 and recorded there,
// never stubbed here.
func reservationProblem(err error) *format.Error {
	if _, ok := live.AsNamespaceError(err); ok {
		return namespaceProblem(err)
	}
	if checkpoint := asCheckpointProblem(err); checkpoint != nil {
		return checkpoint
	}
	var fe *format.Error
	if !errors.As(err, &fe) {
		return sdkProblem(err)
	}
	switch fe.Code {
	case format.CodeAccessPolicyUnsupported:
		return problem(format.CodeAccessPolicyUnsupported, "creator-only access policy is not proved")
	case format.CodeFormatInvalid:
		// The codec and header-invariant arms carry Rust-verbatim
		// fixed details from the producer ("reservation record is
		// malformed" / "reservation state is inconsistent"); preserve
		// them. A bare FormatInvalid class is the output bootstrap
		// fold (Rust reservation_file::Error::Output).
		if fe.Detail == "" {
			return outputProblem(err)
		}
		return fe
	case format.CodeConflict:
		// The header-changed and length-changed arms carry
		// Rust-verbatim fixed details ("reservation record changed" /
		// "reservation length changed"); preserve them.
		return fe
	default:
		return sdkProblem(err)
	}
}

// replacementProblem maps one replacement failure (Rust
// Problem::replacement: Namespace, Sdk, Output, SameIdentity,
// ContentChanged).
func replacementProblem(err error) *format.Error {
	if _, ok := live.AsNamespaceError(err); ok {
		return namespaceProblem(err)
	}
	var fe *format.Error
	if !errors.As(err, &fe) {
		return sdkProblem(err)
	}
	switch fe.Code {
	case format.CodeAccessPolicyUnsupported:
		return problem(format.CodeAccessPolicyUnsupported, "creator-only access policy is not proved")
	case format.CodeFormatInvalid:
		return outputProblem(err)
	case format.CodeConflict:
		// The same-identity and content-changed arms carry
		// Rust-verbatim fixed details from the producer ("replacement
		// source and destination identities match" / "replacement
		// destination content changed"); preserve them.
		return fe
	default:
		return sdkProblem(err)
	}
}

// mainProblem maps one main-file failure (Rust Problem::main:
// Namespace, Sdk, Output, Reservation, Checkpoint, PreviousLinkCount,
// ReservationLinkCount). The Checkpoint arm is recorded with
// 4-10/4-11, the Gc arm is Windows-only Phase 2, and the Injected arm
// is test-only (the Go crash harness maps it at the fault boundary);
// none are stubbed here.
func mainProblem(err error) *format.Error {
	if _, ok := live.AsNamespaceError(err); ok {
		return namespaceProblem(err)
	}
	if checkpoint := asCheckpointProblem(err); checkpoint != nil {
		return checkpoint
	}
	var fe *format.Error
	if !errors.As(err, &fe) {
		return sdkProblem(err)
	}
	switch fe.Code {
	case format.CodeAccessPolicyUnsupported:
		return problem(format.CodeAccessPolicyUnsupported, "creator-only access policy is not proved")
	case format.CodeFormatInvalid:
		return outputProblem(err)
	case format.CodeConflict:
		// The reservation arms carry Rust-verbatim fixed details;
		// preserve them.
		return fe
	case format.CodeCleanupConflict:
		// The retired-link arms carry Rust-verbatim fixed details
		// ("retired previous destination still has a link" /
		// "retired reservation still has a link"); preserve them.
		return fe
	default:
		return sdkProblem(err)
	}
}

// sdkProblem maps one SDK operation failure (Rust Problem::sdk): the
// error code is preserved and the detail is the fixed Rust string; Go
// carries no os_code. Errors that are neither namespace errors nor
// typed SDK errors are unreachable in the ported machine; they fold to
// the Io class like an unclassified syscall failure.
func sdkProblem(err error) *format.Error {
	var fe *format.Error
	if !errors.As(err, &fe) {
		return problem(format.CodeIO, "publication SDK operation failed")
	}
	return problem(fe.Code, "publication SDK operation failed")
}
